### PR TITLE
Release v0.4.0

### PR DESCRIPTION
**Pull request by pmelange**
Update the changelog
Bump the version to 0.4.0
Update the list of supported distros to include OpenWrt

### CODE REVIEW

**Comment by pmelange** [comment]
setup.py will need to be changed again to get it ready for 0.4.1-dev1 (or 0.5.0-dev1).  Let me know what is preferred.

@RalfJung @mitar

**Comment by mitar** [comment]
Let's go with `0.4.1-dev1` after the release and we can always make a bigger version increment afterwards, I think.

Thanks for making it!

**Comment by pmelange** [comment]
If it's looking good enough for the release, then could someone merge and then also tag?  Then I will prepare another PR for 0.4.1-dev1

**Comment by RalfJung** [comment]
Please wait for me to take a look at the changelog, I should get to it over the weekend. It's been years since the last release, no need to rush. :)

File: HISTORY.rst [hunk 1]
```diff
@@ -1,31 +1,49 @@
-vNEXT
+v0.4.0, 2023-Aug-17
 -----
 
+* The Linux kernel which is supported (>= 5.10.152 5.15.76, 6.0.6, 6.1+)
```
Comment (RalfJung) [inline HISTORY.rst:4]:
```suggestion
* The following Linux kernel versions are supported: >= 5.10.152, 5.15.76, 6.0.6, 6.1+
```

File: HISTORY.rst [hunk 2]
```diff
@@ -1,31 +1,49 @@
...
+* Improved documentation and automated checks
```
Comment (RalfJung) [inline HISTORY.rst:5]:
```suggestion
* Improved documentation and automated checks.
```

File: HISTORY.rst [hunk 3]
```diff
@@ -1,31 +1,49 @@
...
 * Python 3 support.
   Python 2 is no longer supported.
   **NOTE:** If you are running the broker in a Python virtualenv you will have to
   rebuild the virtualenv using a Python 3 interpreter.
   This can be achieved by deleting the old virtualenv folder and recreating it.
   For recreating the virtualenv you can refer to the "Installation" section of
   the broker `installation chapter in the Tunneldigger documentation`_.
+* Added cmake buildsystem to the client.
+* Removed NAT implementation.
```
Comment (RalfJung) [inline HISTORY.rst:14]:
```suggestion
* Removed NAT-based handling of many client tunnels on the same server port.
  We now rely on the kernel properly distinguishing those UDP sockets.
```

File: HISTORY.rst [hunk 4]
```diff
@@ -1,31 +1,49 @@
...
   when the broker detects that it runs on an affected kernel, it reports maximal
   usage to old clients -- so if the client does usage-based selection, it will
   pick another broker.
-* Added cmake buildsystem to the client.
-* Added option `connection_rate_limit` to configure the delay between two
-  clients connecting.  Default is 10 (seconds).
-* Added option `pmtu`, defaults to 0 (auto-discovery). If set to a non-zero
-  value, this disables PMTU discovery and replying to client's PMTU discovery
-  probes.
+* Improved PMTU discovery to shortly after tunnel creation.
```
Comment (RalfJung) [inline HISTORY.rst:23]:
```suggestion
* Improved PMTU discovery to complete more quickly after tunnel creation.
```

File: HISTORY.rst [hunk 5]
```diff
@@ -1,31 +1,49 @@
...
+* l2tp_broker.cfg confguration file changes:
+
+  * Added option ``connection_rate_limit`` to configure the delay between two
+    clients connecting.  Default is 10 (seconds).  Valid values include whole
+    seconds (10) or fractions of a second (0.2).  A value of 0 disables this
+    feature.
+  * Added options ``connection_rate_limit_per_ip_count`` and 
+    ``connection_rate_limit_per_ip_time`` to configure a connection rate limit
+    per client IP address.  If either value is 0, the functionality is disabled.
+    The default values are 0.
```
Comment (RalfJung) [inline HISTORY.rst:37]:
```suggestion
    The default values are 0. This is applied per broker port.
```

File: HISTORY.rst [hunk 6]
```diff
@@ -1,31 +1,49 @@
...
+  * Added hook ``broker.connection-rate-limit`` which is called when the connection
+    rate per IP address limit is exceeded
```
Comment (RalfJung) [inline HISTORY.rst:39]:
```suggestion
    rate per IP address limit is exceeded.
```

File: HISTORY.rst [hunk 7]
```diff
@@ -1,31 +1,49 @@
...
+  * Added option ``pmtu``, defaults to 0 (auto-discovery). If set to a non-zero
+    value, this disables PMTU discovery and replying to client's PMTU discovery
+    probes.
+  * Removed option ``filename``. All log message are sent to STDOUT or STDERR
+  * Removed option ``port_base``.
```
Comment (RalfJung) [inline HISTORY.rst:44]:
```suggestion
  * Removed option ``port_base``. No longer needed without the NAT-based approach.
```

File: HISTORY.rst [hunk 8]
```diff
@@ -1,31 +1,49 @@
...
-* Removed dependency on Netfilter.
-* Removed NAT hack.  However, this means we now require a `very recent kernel`_
-  (>= 5.2.17, 5.3.1, 5.4+).
+* Improved and more detailed logging
+* l2tp_broker.cfg confguration file changes:
+
+  * Added option ``connection_rate_limit`` to configure the delay between two
+    clients connecting.  Default is 10 (seconds).  Valid values include whole
+    seconds (10) or fractions of a second (0.2).  A value of 0 disables this
+    feature.
```
Comment (RalfJung) [inline HISTORY.rst:33]:
```suggestion
  * Added option ``connection_rate_limit`` to configure the delay between two
    clients connecting.  Default is 10 (seconds), which is the limit applied by previous versions.
    Valid values include whole seconds (10) or fractions of a second (0.2).
    A value of 0 disables this feature. This is applied per broker port.
```

File: HISTORY.rst [hunk 9]
```diff
@@ -1,31 +1,49 @@
...
+    per client IP address.  If either value is 0, the functionality is disabled.
+    The default values are 0.
+  * Added hook ``broker.connection-rate-limit`` which is called when the connection
+    rate per IP address limit is exceeded
+  * Added option ``pmtu``, defaults to 0 (auto-discovery). If set to a non-zero
+    value, this disables PMTU discovery and replying to client's PMTU discovery
+    probes.
+  * Removed option ``filename``. All log message are sent to STDOUT or STDERR
+  * Removed option ``port_base``.
+  * Removed option ``namespace``.
```
Comment (RalfJung) [inline HISTORY.rst:45]:
```suggestion
  * Removed option ``namespace``. No longer needed without the NAT-based approach.
```

File: HISTORY.rst [hunk 10]
```diff
@@ -1,31 +1,49 @@
...
+  * Removed option ``check_modules``.
```
Comment (RalfJung) [inline HISTORY.rst:46]:
```suggestion
  * Removed option ``check_modules``. It already stopped doing anything since v0.2.0.
```

**Review by RalfJung — COMMENTED** [review]
Thanks for going through all the changes! I just have some editorial suggestions.

File: HISTORY.rst (hunk 1 above)
Comment (pmelange) [reply HISTORY.rst]: done

File: HISTORY.rst (hunk 2 above)
Comment (pmelange) [reply HISTORY.rst]: done

File: HISTORY.rst (hunk 3 above)
Comment (pmelange) [reply HISTORY.rst]: Changed it so "we" is not used.  Instead "The broker now relies on the kernel properly distinguishing those UDP sockets." instead.  Done.

File: HISTORY.rst (hunk 4 above)
Comment (pmelange) [reply HISTORY.rst]: done

File: HISTORY.rst (hunk 5 above)
Comment (pmelange) [reply HISTORY.rst]: done

File: HISTORY.rst (hunk 6 above)
Comment (pmelange) [reply HISTORY.rst]: done

File: HISTORY.rst (hunk 7 above)
Comment (pmelange) [reply HISTORY.rst]: done

File: HISTORY.rst (hunk 8 above)
Comment (pmelange) [reply HISTORY.rst]: Done

File: HISTORY.rst (hunk 10 above)
Comment (pmelange) [reply HISTORY.rst]: Reworded to "No longer implemented since v0.2.0"

File: HISTORY.rst (hunk 9 above)
Comment (pmelange) [reply HISTORY.rst]: done

**Comment by pmelange** [comment]
If it's looking good enough for the release, then could someone merge and then also tag?  Then I will prepare another PR for 0.4.1-dev1

> Thanks for going through all the changes! I just have some editorial suggestions.

Thanks for the suggestions.  Most of them I did verbatim.  A couple I reworded.

**Comment by PolynomialDivision** [comment]
ping @RalfJung ;) Would be great to get this merged to finally have a complete working version with openwrt. Thanks a lot for your work.

**Comment by RalfJung** [comment]
Merged the PR and pushed the tag. :)

**Comment by pmelange** [comment]
> Merged the PR and pushed the tag. :)
Thank you @RalfJung

**Pull request merged**